File: generated_code/GeminiPro/Java/063_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splay operation. Moves the node with the given key (or the last
     * accessed node if the key is not found) to the root of the tree.
     * This implementation uses the top-down splaying technique.
     *
     * @param key The key to splay around.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Create a dummy node to simplify linking
        Node dummy = new Node(0);
        Node leftTree = dummy;
        Node rightTree = dummy;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig (Left-Left) case
                if (key < current.left.key) {
                    current = rightRotate(current);
                    if (current.left == null) {
                        break;
                    }
                }
                // Link current to the right tree
                rightTree.left = current;
                rightTree = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig (Right-Right) case
                if (key > current.right.key) {
                    current = leftRotate(current);
                    if (current.right == null) {
                        break;
                    }
                }
                // Link current to the left tree
                leftTree.right = current;
                leftTree = current;
                current = current.right;
            } else { // key == current.key
                break;
            }
        }

        // Reassemble the tree
        leftTree.right = current.left;
        rightTree.left = current.right;
        current.left = dummy.right;
        current.right = dummy.left;
        root = current;
    }

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        // After splaying, the root is the closest node to the key.
        // We check if it's an exact match.
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree. Duplicates are not allowed.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. If the key is present, it becomes the root.
        // If not, the node that would be its parent becomes the root.
        splay(key);

        // If the key is already at the root, it's a duplicate; do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the key to the root.
        splay(key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // Now, the node to delete is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Make the left subtree the main tree.
            root = leftSubtree;
            
            // Splay for the key we just deleted in the new tree (the left subtree).
            // This will bring the largest element in the left subtree (the predecessor
            // of the deleted key) to the root.
            splay(key);
            
            // The new root (max of the left subtree) is guaranteed to have no right child.
            // Attach the original right subtree as its right child.
            root.right = rightSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/063_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Python class implementing a Splay Tree.

    This class provides a dictionary-like set for storing unique integer keys.
    It supports insertion, deletion, and searching operations. A key feature
    is the splaying operation, which moves accessed elements to the root of
    the tree to optimize for future accesses.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def insert(self, key):
        """
        Inserts a key into the Splay Tree.

        If the key already exists, the node with that key is splayed to the root.
        If the key is new, it is inserted as in a standard BST, and the new
        node is then splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        if self.root is None:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current is not None:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay it and return
                self._splay(current)
                return

        new_node = self._Node(key)
        new_node.parent = parent
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        self._splay(new_node)

    def search(self, key):
        """
        Searches for a key in the Splay Tree.

        If the key is found, the corresponding node is splayed to the root
        and the method returns True. If the key is not found, the last
        accessed node (the would-be parent) is splayed to the root and
        the method returns False.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        current = self.root
        last_visited = None
        while current is not None:
            last_visited = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                self._splay(current)
                return True
        
        # Key not found, splay the last visited node (parent)
        if last_visited:
            self._splay(last_visited)
        return False

    def delete(self, key):
        """
        Deletes a key from the Splay Tree.

        The node with the given key is first splayed to the root. If it exists,
        it is removed. The tree is then reformed by joining the left and right
        subtrees. The largest element of the left subtree becomes the new root.

        Args:
            key (int): The integer key to delete.
        """
        if not self.root:
            return

        # Splay the node with the key (or its parent) to the root
        self.search(key)

        # If the key is not at the root, it wasn't in the tree
        if self.root.key != key:
            return

        # Now the node to delete is the root
        left_subtree = self.root.left
        right_subtree = self.root.right

        if left_subtree is None:
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif right_subtree is None:
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Disconnect the left subtree from the old root
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Temporarily make the left subtree the main tree to splay within it
            temp_root = self.root
            self.root = left_subtree
            self._splay(max_node)
            
            # After splaying, max_node is the new root of the (modified) left subtree.
            # It has no right child. We attach the original right subtree here.
            self.root.right = right_subtree
            right_subtree.parent = self.root

    # ----------------- Private Helper Methods -----------------

    def _left_rotate(self, x):
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node):
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if grandparent is None:
                # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            else:
                if node == parent.left:
                    if parent == grandparent.left:
                        # Zig-Zig case (left-left)
                        self._right_rotate(grandparent)
                        self._right_rotate(parent)
                    else:
                        # Zig-Zag case (right-left)
                        self._right_rotate(parent)
                        self._left_rotate(grandparent)
                else:
                    if parent == grandparent.right:
                        # Zig-Zig case (right-right)
                        self._left_rotate(grandparent)
                        self._left_rotate(parent)
                    else:
                        # Zig-Zag case (left-right)
                        self._left_rotate(parent)
                        self._right_rotate(grandparent)